### PR TITLE
Improve test coverage for impl edge cases, parser API, and complexity classes

### PR DESCRIPTION
**Pull request by DominicBurkart**
## Summary\n\nThis PR adds tests targeting three undertested areas of the codebase:\n\n### 1. Integration tests for the `m!` macro (`tests/tests/macro_edge_cases.rs`)\n- **Edge cases**: empty range, single-element range, inclusive single-element range\n- **Fold**: sum over a range, fold over empty stream returns init value\n- **Composition**: chained map+filter, filter-rejects-all, filter-accepts-all, chained maps, map identity\n- **Combinatorics**: combinations/permutations of single element\n- **Scale**: large range (1000 elements) correctness check\n- **select_all**: single-stream variant\n\n### 2. Parser API tests (`marigold-grammar/tests/e2e_parse.rs`)\n- **Happy paths**: Exercises `marigold_parse()` directly with all major language constructs (range, map, filter, permutations, combinations, fold, keep_first_n, select_all, fn declarations, write_file, chained operations, inclusive range)\n- **Sad paths**: Empty input, garbage input, missing terminal, unclosed parenthesis\n- **Analysis API**: `marigold_analyze()` for stream count, cardinality preservation through map, cardinality reduction through filter\n\n### 3. Deterministic complexity class tests (`marigold-grammar/tests/complexity_unit.rs`)\n- **ComplexityClass ordering**: full ordering chain verification, polynomial/combinatorial/permutational sub-ordering by degree/k\n- **Display/FromStr roundtrip**: all 10 variants\n- **ExactComplexity**: new/empty, add_work, simplified picks highest class, merge combines coefficients, Display/FromStr roundtrip\n- **Cardinality**: Display, max semantics (exact vs exact, exact vs unknown), ordering (exact < unknown, exact ordering by value)\n\n### 4. Extended cardinality integration tests (`marigold-grammar/tests/e2e_cardinality_extended.rs`)\n- Edge cases: empty range (cardinality 0), single element, permutations_with_replacement\n- Inline analysis: fold always produces cardinality 1, filter produces bounded, program cardinality = max of streams\n\n## Key invariants validated\n- `range(0, 0)` yields an empty stream\n- `fold` over an empty stream returns the init value\n- `filter` that rejects/accepts everything produces empty/full stream\n- `ComplexityClass` ordering is total and monotonic\n- `Display`/`FromStr` roundtrips are lossless for all complexity types\n- `Cardinality::max` returns `Unknown` when either side is unknown\n- Program cardinality equals the max of all stream cardinalities\n\n## Test plan\n- [ ] `cargo test` passes (no features)\n- [ ] `cargo test --features tokio` passes\n- [ ] `cargo test --all-features` passes"

### CODE REVIEW

File: tests/tests/macro_edge_cases.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+#[tokio::test]
+async fn permutations_of_single_element() {
+    let r = m!(
+        range(0, 1)
+            .permutations(1)
+            .return
+    )
+    .await
+    .collect::<Vec<_>>()
```
Comment (DominicBurkart) [inline tests/tests/macro_edge_cases.rs]: `map_identity` and `filter_accepts_all` are testing that the stream infrastructure passes elements through unchanged. These are essentially testing `StreamExt::map` and `StreamExt::filter` with trivial functions, not marigold-specific logic. The `m!` macro's job is code generation -- these tests don't exercise any interesting macro expansion path beyond what `single_element_range` already covers.

Similarly, `large_range_count` just verifies that `range(0, 1000)` produces 1000 elements. This is testing the range iterator, not a macro edge case. Consider whether these are worth the CI time.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: marigold-grammar/tests/e2e_cardinality_extended.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+hundreds.return
+"#,
+    );
+    assert_eq!(result.streams.len(), 2);
+    assert_eq!(
+        result.program_cardinality,
+        Cardinality::Exact(BigUint::from(100u64)),
+        "program cardinality should be max of all streams"
+    );
+}
```
Comment (DominicBurkart) [inline marigold-grammar/tests/e2e_cardinality_extended.rs]: This test asserts `program_cardinality` equals the max of streams, but it only tests the case where both streams have `Exact` cardinality. The more interesting edge case -- program cardinality when one stream is `Bounded` or `Unknown` -- is missing. That would actually exercise the `Cardinality::max` logic with mixed variants in a realistic context.

---
_Generated by [Claude Code](https://claude.ai/code)_
